### PR TITLE
Preserve Txn revert reason

### PR DESCRIPTION
**Pull request by ppedemon**
RSK is not notifying the revert reason when a transaction invoked by RPC is reverted. This PR decodes and adds the revert reason to the exception thrown in such case.

### CODE REVIEW

**Comment by pmprete** [comment]
A much needed message when developing. 
Great contribution @ppedemon ppedemon!

File: rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java [hunk 1]
```diff
@@ -227,6 +240,25 @@ private ProgramResult callConstant(Web3.CallArguments args, Block executionBlock
...
+        String value = Hex.toHexString(res.getHReturn());
+        if (!value.startsWith("08c379a0")) {
+            return Optional.empty();
+        }
+
+        List<TypeReference<Type>> revertReasonTypes =
+                Collections.singletonList(TypeReference.create((Class<Type>) AbiTypes.getType("string")));
+        String encodedRevertReason = value.substring(8);
+        List<Type> decoded = FunctionReturnDecoder.decode(encodedRevertReason, revertReasonTypes);
+        return Optional.of(decoded.get(0).getValue().toString());
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java:259]:
```suggestion
        return decoded.isEmpty() ? Optional.empty() : Optional.of(decoded.get(0).getValue().toString());
```

**Review by Vovchyk — CHANGES_REQUESTED** [review]

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

File: rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java [hunk 2]
```diff
@@ -227,6 +240,25 @@ private ProgramResult callConstant(Web3.CallArguments args, Block executionBlock
...
 
+    @SuppressWarnings("unchecked")
+    private Optional<String> decodeRevertReason(ProgramResult res) {
+        byte[] bytes = res.getHReturn();
+        if (bytes == null || bytes.length == 0) {
+            return Optional.empty();
+        }
+
+        String value = Hex.toHexString(res.getHReturn());
+        if (!value.startsWith("08c379a0")) {
```
Comment (patogallaiovlabs) [inline rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java:251]: Magic value? Add comment and Constant.

File: rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java [hunk 3]
```diff
@@ -227,6 +240,25 @@ private ProgramResult callConstant(Web3.CallArguments args, Block executionBlock
...
+            return Optional.empty();
+        }
+
+        List<TypeReference<Type>> revertReasonTypes =
+                Collections.singletonList(TypeReference.create((Class<Type>) AbiTypes.getType("string")));
+        String encodedRevertReason = value.substring(8);
```
Comment (patogallaiovlabs) [inline rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java:257]: Extract these lines to a method and add docs (at least some doc).

File: rskj-core/src/test/java/co/rsk/rpc/modules/eth/EthModuleTest.java [hunk 4]
```diff
@@ -78,6 +80,50 @@ public void callSmokeTest() {
...
+        ExecutionBlockRetriever retriever = mock(ExecutionBlockRetriever.class);
+        when(retriever.getExecutionBlock_workaround("latest"))
+                .thenReturn(blockResult);
+        when(blockResult.getBlock()).thenReturn(block);
+
+        byte[] hreturn = Hex.decode(
+                "08c379a000000000000000000000000000000000000000000000000000000000" +
+                        "0000002000000000000000000000000000000000000000000000000000000000" +
+                        "0000000f6465706f73697420746f6f2062696700000000000000000000000000" +
+                        "00000000");
```
Comment (patogallaiovlabs) [inline rskj-core/src/test/java/co/rsk/rpc/modules/eth/EthModuleTest.java:97]: What does this value means? Add some comments. We could check other possible values (?)

**Review by patogallaiovlabs — CHANGES_REQUESTED** [review]
Looks good. I suggested some comments to make it more clear.

**Comment by aeidelman** [comment]
pipeline: run

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Comment by fedejinich** [comment]
lgtm

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Comment by patogallaiovlabs** [comment]
pipeline:run

File: rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java [hunk 5]
```diff
@@ -58,6 +58,9 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger("web3");
 
+    private static final CallTransaction.Function ERROR_ABI_FUNCTION = CallTransaction.Function.fromSignature("Error", "string");
+    private static final byte[] ERROR_ABI_FUNCTION_SIGNATURE = ERROR_ABI_FUNCTION.encodeSignature(); //08c379a0
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java:62]: Any real transaction executed in a code test (maybe using DSL), to test this signature?

Comment (patogallaiovlabs) [reply rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModule.java:62]: Done!

**Review by ajlopezrsk — APPROVED** [review]

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Comment by patogallaiovlabs** [comment]
pipeline:run

**Review by patogallaiovlabs — APPROVED** [review]

**Pull request merged**